planning/MartenV4-WastazComments.md

```diff
@@ -0,0 +1,174 @@
+now. But if I had to implement this again knowing what I do today I would 100% certainly implement publishing to the message bus as an async projection.
+
+Running projections off the message bus was also a thing that we tried doing for a while. It has some good parts, but also some bad parts. The main issue
+is of course that you suddenly end up having to handle things like out-of-order messages, duplicated messages, not being sure of where you are in the stream
+etc. Its quite a lot to keep in mind, but for some cases its quite useful. But complexity immediately rises, especially in writing the projections, and lets not
+even talk about event replay. I saw that you were discussing allowing projections from a message bus, and it would be a great feature to have but be prepared
+that in order to support the same type of projections as you can do today is gonna require quite a bit of infrastructure. But if you can pull it off, Id certainly be happy! ;)
+
+
+# Immutable objects
+
+Honestly, you knew this were coming. But hear me out.
+
+## Immutable events
+Events should be immutable, thats just a basic fact of eventsourcing. But due to serialization woes etc the common implementation of events are with 
+`public string Fruit { get; set; }` because thats just life in C# world. Now, this "gentleman immutability" "works" but it is...not great?
+
+### It aint immutable
+Sure, if you go assign stuff to it and end up in trouble because of it thats your own fault. But I am also a firm believer in the principle of least surprise, if you shouldnt 
+set a property there shouldnt be a setter for it. If a method shouldnt be in the public interface of a class then make it private or internal.
+
+### F# support
+Mutable bags of data like this is highly non-idiomatic in F#, and key to providing good F# support is going to be in part allowing for better support for 
+immutable object. The normal modus operandi there would be to implement an event as a record type (which in turn is probably wrapped in a DU, but lets table that discussion for right now)
+
+```fsharp
+type MyEvent = {
+  fruit : string
+  numberOfFruitsBought : int
+}
+```
+
+This event is immutable. By adding the `[<CLIMutable>]` attribute on it we can get public setters generated on it as well that cant be used from F#, solving some
+issues with immutability and serialization but at the cost of your immutable object...not actually being immutable anymore.
+
+## Immutable aggregates
+
+I dont necessarily think that this should be the default way of working because there is certainly a cost to pay for allocating a lot of objects. But in F#-code or more functional-style code it would 
+be a lot nicer to be able to not mutate the aggregate class but instead return a new instance of the aggregate with the current event applied to it. And honestly I think that it wouldnt be that hard to
+support this mode of operation as well as the mutable way of working.
+
+
+# Nullable reference types
```

Comment:
Looking at https://devblogs.microsoft.com/dotnet/try-out-nullable-reference-types/ I presume huge maintainance burden. MS went a bit nutts with that. I'd love to use the Nullable Reference Type but the number of permutations that might be generated by the users might be overwhelming... 

Some food for thought from the Jon Skeet that was "fighting" with that in NodaTime:
- https://codeblog.jonskeet.uk/2018/04/21/first-steps-with-nullable-reference-types/
- https://codeblog.jonskeet.uk/2019/02/10/nullableattribute-and-c-8/
- https://www.youtube.com/watch?v=1tpyAQZFlZY